### ISSUE TITLE
StackOverflowError when printing self-referential struct

### ISSUE THREAD

**Issue by HFriberg** [JuliaLang/julia#33392]
I expected the following code
```
mutable struct node_t
    prev::node_t
    next::node_t
end

x = Vector{node_t}(undef,3);
show(x)  # StackOverflowError
```

to produce the output
```
node_t[#undef, #undef, #undef]
```
but instead it recursed forever until the stack was full. 

**Bonus info:** Adding an unused(?) inner constructor:
```
mutable struct node_t
    prev::node_t
    next::node_t
    node_t() = new()    # not used anywhere??
end

x = Vector{node_t}(undef,3);
show(x)
```
makes it behave as expected.

**Comment by jpsamaroo** [comment]
I can't reproduce this locally, what version of Julia are you on?

```
julia> mutable struct node_t
           prev::node_t
           next::node_t
       end

julia> x = Vector{node_t}(undef,3);

julia> show(x)  # StackOverflowError
node_t[#undef, #undef, #undef]
julia> versioninfo()
Julia Version 1.3.0-alpha.105
Commit 6527de0c0f (2019-08-12 22:31 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, haswell)
Environment:
  JULIA_REVISE_POLL = 1
  JULIA_LOAD_PATH = @dev:@:@v#.#:@stdlib
  JULIA_PKG3_PRECOMPILE = 1
```

**Comment by JeffBezanson** [comment]
I believe this is a dup of #31064 which is fixed in 1.3.

**Comment by HFriberg** [comment]
@JeffBezanson is correct.
